MoreLinq.Test/SliceTest.cs

```diff
@@ -111,19 +112,34 @@ public void TestSliceLongerThanSequence()
         }
 
         /// <summary>
-        /// Verify that slice is optimized for <see cref="IList{T}"/> implementations and does not
+        /// Verify that slice is optimized for <see cref="IList{T}"/> and <see cref="IReadOnlyList{T}"/> implementations and does not
         /// unnecessarily traverse items outside of the slice region.
         /// </summary>
         [Test]
-        public void TestSliceListOptimization()
+        [TestCaseSource(nameof(TestSliceOptimizationCases))]
```

Comment:
I was thinking of something much simpler, like:

```c#
        [TestCase(false)]
        [TestCase(true)]
        public void TestSliceOptimization(bool readOnly)
        {
            const int sliceStart = 4;
            const int sliceCount = 3;
            var sequence = readOnly
                         ? new UnenumerableReadOnlyList<int>(new[] { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 })
                         : (IEnumerable<int>) new UnenumerableList<int> { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 };

            var result = sequence.Slice(sliceStart, sliceCount);

            Assert.AreEqual(sliceCount, result.Count());
            CollectionAssert.AreEqual(Enumerable.Range(5, sliceCount), result);
        }
```

Reply:
Works for me.